src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs

```diff
@@ -433,23 +453,49 @@ private void PopulateReferenceIdDictionary(IEnumerable<EntryComponent> bundleEnt
         {
             foreach (EntryComponent entry in bundleEntries)
             {
-                if (entry.Request.Method != HTTPVerb.POST)
-                {
-                    continue;
-                }
+                var versionId = string.Empty;
 
-                // We've already come across this ID
-                if (!string.IsNullOrWhiteSpace(entry.FullUrl) && !idDictionary.ContainsKey(entry.FullUrl))
+                if (!string.IsNullOrWhiteSpace(entry.FullUrl))
                 {
-                    // This id is new to us
-                    var insertId = _resourceIdProvider.Create();
-                    entry.Resource.Id = insertId;
+                    if (entry.FullUrl.Contains("/_history/", StringComparison.OrdinalIgnoreCase))
+                    {
+                        string[] versionedReference = entry.FullUrl.Split("/_history/");
+                        versionId = versionedReference[1];
+                    }
 
-                    idDictionary.Add(entry.FullUrl, (insertId, entry.Resource.TypeName));
+                    var requestUrl = entry.Request?.Url;
+
+                    if (entry.Request.Method == HTTPVerb.POST)
+                    {
+                        // We've already come across this ID
+                        if (!idDictionary.ContainsKey(entry.FullUrl))
+                        {
+                            // This id is new to us
+                            var insertId = _resourceIdProvider.Create();
+                            entry.Resource.Id = insertId;
+
+                            idDictionary.Add(entry.FullUrl, (insertId, entry.Resource.TypeName));
+                        }
+                    }
+                    else if (!string.IsNullOrWhiteSpace(versionId) && requestUrl != null)
+                    {
+                        // Extract resourceId from requestUrl.
+                        var resourceId = IsRequestUrlContainsResourceId(requestUrl) ? requestUrl.Split("/")[1] : null;
+
+                        if (resourceId != null)
+                        {
+                            idDictionary.Add(entry.FullUrl, (resourceId, entry.Resource.TypeName));
+                        }
+                    }
                 }
             }
         }
 
+        private static bool IsRequestUrlContainsResourceId(string requestUrl)
+        {
+            return requestUrl.Contains("/", StringComparison.Ordinal) && !requestUrl.Contains("?", StringComparison.Ordinal);
```

Comment:
Wouldn't a `requestUrl` of `/Patient` satisfy this check?

Reply:
The implementation has changed since the last time the code has been reviewd, This code flow doesnot exists in the current code flow.